File: windows/Wetterkurve.Core/SettingsStore.cs
using System.Text.Json;

namespace Wetterkurve;

public static class SettingsStore
{
    public static string DefaultDirectory => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "Wetterkurve");

    public static string DefaultPath => Path.Combine(DefaultDirectory, "settings.json");

    public static AppState Load(string? path = null)
    {
        path ??= DefaultPath;
        try
        {
            if (!File.Exists(path))
                return DefaultState();
            var json = File.ReadAllText(path);
            using var document = JsonDocument.Parse(json);
            var root = document.RootElement;
            var locationsJson = root.TryGetProperty("locations", out var locations)
                ? locations.GetRawText()
                : "[]";
            var parsed = WeatherService.ParseLocations(locationsJson, [WeatherService.DefaultLocation]);
            var active = 0;
            if (root.TryGetProperty("activeLocation", out var activeElement))
                active = activeElement.GetInt32();
            double? indicatorLeft = root.TryGetProperty("indicatorLeft", out var leftElement) &&
                leftElement.ValueKind == JsonValueKind.Number
                ? leftElement.GetDouble()
                : null;
            double? indicatorTop = root.TryGetProperty("indicatorTop", out var topElement) &&
                topElement.ValueKind == JsonValueKind.Number
                ? topElement.GetDouble()
                : null;
            return new AppState
            {
                Locations = parsed,
                ActiveLocation = Math.Clamp(active, 0, parsed.Count - 1),
                IndicatorLeft = indicatorLeft,
                IndicatorTop = indicatorTop,
            };
        }
        catch
        {
            return DefaultState();
        }
    }

    public static void Save(AppState state, string? path = null)
    {
        path ??= DefaultPath;
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        var json = JsonSerializer.Serialize(new
        {
            locations = JsonSerializer.Deserialize<JsonElement>(WeatherService.SerializeLocations(state.Locations)),
            activeLocation = state.ActiveLocation,
            indicatorLeft = state.IndicatorLeft,
            indicatorTop = state.IndicatorTop,
        });
        File.WriteAllText(path, json);
    }

    public static AppState DefaultState() => new()
    {
        Locations = [WeatherService.DefaultLocation],
        ActiveLocation = 0,
    };
}

File: android/core/src/main/kotlin/de/wean/wetterkurve/SettingsStore.kt
package de.wean.wetterkurve

import kotlinx.serialization.json.Json
import kotlin.math.min

object SettingsStore {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    fun defaultState(): AppState = AppState(
        locations = listOf(WeatherService.defaultLocation),
        activeLocation = 0,
    )

    fun parse(value: String?): AppState {
        return try {
            val parsed = json.decodeFromString<AppStateJson>(value ?: "")
            val locationsJson = json.encodeToString(
                kotlinx.serialization.builtins.ListSerializer(WeatherLocationJson.serializer()),
                parsed.locations,
            )
            val locations = WeatherService.parseLocations(
                locationsJson,
                listOf(WeatherService.defaultLocation),
            )
            AppState(
                locations = locations,
                activeLocation = parsed.activeLocation.coerceIn(0, locations.lastIndex),
                showClouds = parsed.showClouds,
                showWind = parsed.showWind,
            )
        } catch (_: Exception) {
            defaultState()
        }
    }

    fun serialize(state: AppState): String {
        val locations = WeatherService.parseLocations(
            WeatherService.serializeLocations(state.locations),
            listOf(WeatherService.defaultLocation),
        )
        val active = if (locations.isEmpty()) 0 else min(state.activeLocation.coerceAtLeast(0), locations.lastIndex)
        return json.encodeToString(
            AppStateJson.serializer(),
            AppStateJson(
                locations = json.decodeFromString(
                    kotlinx.serialization.builtins.ListSerializer(WeatherLocationJson.serializer()),
                    WeatherService.serializeLocations(locations),
                ),
                activeLocation = active,
                showClouds = state.showClouds,
                showWind = state.showWind,
            ),
        )
    }
}
